src/header/common/content_disposition.rs

```diff
@@ -0,0 +1,384 @@
+                    _ => DispositionType::Ext(disposition),
+                },
+                parameters: Vec::new(),
+            };
+
+            for section in sections {
+                let mut parts = section.split('=');
+
+                let key = if let Some(key) = parts.next() {
+                    key.trim().to_ascii_lowercase()
+                } else {
+                    return Err(::Error::Header);
+                };
+
+                let val = if let Some(val) = parts.next() {
+                    val.trim()
+                } else {
+                    return Err(::Error::Header);
+                };
+
+                cd.parameters.push(
+                    match &*key {
+                        "filename" => DispositionParam::Filename(
+                            Charset::Ext("UTF-8".to_owned()), None, unquote_value(val).into_bytes()),
+                        "filename*" => {
+                            let (charset, opt_language, value) = try!(parse_ext_value(val));
+                            DispositionParam::Filename(charset, opt_language, value)
+                        },
+                        _ => DispositionParam::Ext(key, unquote_value(val)),
+                    }
+                );
+            }
+
+            Ok(cd)
+        })
+    }
+}
+
+impl HeaderFormat for ContentDisposition {
+    #[inline]
+    fn fmt_header(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self, f)
+    }
+}
+
+impl fmt::Display for ContentDisposition {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.disposition {
+            DispositionType::Inline => try!(write!(f, "inline")),
+            DispositionType::Attachment => try!(write!(f, "attachment")),
+            DispositionType::Ext(ref s) => try!(write!(f, "{}", s)),
+        }
+        for param in self.parameters.iter() {
+            match param {
+                &DispositionParam::Filename(ref charset, ref opt_lang, ref bytes) => {
+                    let mut use_simple_format: bool = false;
+                    if opt_lang.is_none() {
+                        if let Charset::Ext(ref ext) = *charset {
+                            if ext.to_ascii_lowercase() == "utf-8" {
+                                use_simple_format = true;
+                            }
+                        }
+                    }
+                    if use_simple_format {
+                        try!(write!(f, "; filename=\"{}\"",
+                                    match String::from_utf8(bytes.clone()) {
+                                        Ok(s) => s,
+                                        Err(_) => return Err(fmt::Error),
+                                    }));
+                    } else {
+                        let langstr = match *opt_lang {
+                            Some(ref lang) => format!("{}",lang),
+                            None => "".to_owned(),
```

Comment:
These conditions could write directly to the Formatter, instead of temporary strings.